File: pipeline/lib/http.js
// Tiny fetch wrapper: timeout, retries, polite delay. Uses global fetch (Node 18+).
'use strict';

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

async function getJSON(url, opts = {}) {
  return get(url, { ...opts, as: 'json' });
}
async function getText(url, opts = {}) {
  return get(url, { ...opts, as: 'text' });
}

async function get(url, opts = {}) {
  const {
    as = 'json',
    headers = {},
    timeout = 60000,
    retries = 2,
    retryDelay = 1500,
  } = opts;
  let lastErr;
  for (let attempt = 0; attempt <= retries; attempt++) {
    const ctrl = new AbortController();
    const t = setTimeout(() => ctrl.abort(), timeout);
    try {
      const res = await fetch(url, {
        headers: { 'User-Agent': 'ResearchAgent/0.1', ...headers },
        signal: ctrl.signal,
        redirect: 'follow',
      });
      clearTimeout(t);
      if (res.status === 429 || res.status >= 500) {
        throw new Error(`HTTP ${res.status}`);
      }
      if (!res.ok) {
        const body = await res.text().catch(() => '');
        throw Object.assign(new Error(`HTTP ${res.status} ${url}`), {
          status: res.status,
          body: body.slice(0, 300),
          fatal: res.status >= 400 && res.status < 500 && res.status !== 429,
        });
      }
      return as === 'json' ? await res.json() : await res.text();
    } catch (e) {
      clearTimeout(t);
      lastErr = e;
      if (e.fatal) break;
      if (attempt < retries) await sleep(retryDelay * (attempt + 1));
    }
  }
  throw lastErr;
}

module.exports = { get, getJSON, getText, sleep };

File: pipeline/lib/http.py
"""Tiny HTTP wrapper: timeout, retries, polite delay (uses requests)."""
import time

import requests

DEFAULT_UA = "ResearchAgent/0.1"


class HttpError(Exception):
    def __init__(self, message, status=None, body=None, fatal=False):
        super().__init__(message)
        self.status = status
        self.body = body
        self.fatal = fatal


def sleep(seconds):
    time.sleep(seconds)


def get(url, as_="json", headers=None, timeout=60, retries=2, retry_delay=1.5):
    """GET with retries. 429/5xx are retried; other 4xx are fatal (no retry)."""
    hdrs = {"User-Agent": DEFAULT_UA}
    if headers:
        hdrs.update(headers)
    last_err = None
    for attempt in range(retries + 1):
        try:
            res = requests.get(url, headers=hdrs, timeout=timeout, allow_redirects=True)
            if res.status_code == 429 or res.status_code >= 500:
                raise HttpError(f"HTTP {res.status_code}", status=res.status_code)
            if not res.ok:
                raise HttpError(
                    f"HTTP {res.status_code} {url}",
                    status=res.status_code,
                    body=res.text[:300],
                    fatal=(400 <= res.status_code < 500 and res.status_code != 429),
                )
            return res.json() if as_ == "json" else res.text
        except HttpError as e:
            last_err = e
            if e.fatal:
                break
            if attempt < retries:
                time.sleep(retry_delay * (attempt + 1))
        except requests.RequestException as e:
            last_err = HttpError(str(e))
            if attempt < retries:
                time.sleep(retry_delay * (attempt + 1))
    raise last_err


def get_json(url, **kw):
    return get(url, as_="json", **kw)


def get_text(url, **kw):
    return get(url, as_="text", **kw)


def download_pdf(url, dest, ua=DEFAULT_UA, timeout=60):
    """Download a PDF to `dest`. Raises if the response isn't a real PDF
    (e.g. a publisher landing/Cloudflare page). Returns byte length."""
    res = requests.get(url, headers={"User-Agent": ua, "Accept": "application/pdf,*/*"},
                       timeout=timeout, allow_redirects=True)
    if not res.ok:
        raise HttpError(f"HTTP {res.status_code}", status=res.status_code)
    data = res.content
    if data[:5] != b"%PDF-":
        raise HttpError("not a PDF (likely landing page)")
    with open(dest, "wb") as f:
        f.write(data)
    return len(data)
